src/core/connections.jl

```diff
@@ -258,10 +261,28 @@ function add_external_param_conn(md::ModelDef, conn::ExternalParameterConnection
     push!(md.external_param_conns, conn)
 end
 
-function set_external_param!(md::ModelDef, name::Symbol, value::ModelParameter)
+function set_external_param!(md::ModelDef, name::Symbol, value::ModelParameter; param_dims = [], num_dims = 0)
```

Comment:
It looks like the two optional parameters can be replaced by one, `comp_name`, allowing `param_dims` and `num_dims` to be computed in the `set_external_param!` rather than by the caller (which ends up propagating two optional parameters through several methods):
```
 param_dims = parameter_dimensions(md, comp_name, param_name)
 num_dims = length(param_dims)
```

Reply:
@rjplevin great, I wasn't sure how to access the information without other inputs.  So this will be a different way to access the same values as below.  Then as you said we can remove optional parameters from everything ... I was hoping we'd be able to do that!

```
param_dims = Mimi.parameter_dimensions(m, component, name)    
num_dims = length(size(value))
```

Reply:
@rjplevin the issue here is that we cannot call `param = external_param(md, name)`, because the parameter in question is currently unconnected and thus not in the list of external parameters.  Instead, the only way to access it's dimensions is to look at the component itself, but we don't want to have to pass the component name through if we can avoid it.  **Assuming that all external parameters must be uniquely named** (can we definitely assume this though?), we could go call `unconnected_params(md)` to get the component name, which adds some work but avoids passing around the component name.   Alternatively we can pass in the component name and keep that one extra optional arg.

Previous `set_leftover_params!`:

```
for (comp_name, param_name) in leftovers
        # check whether we need to set the external parameter
        if ! haskey(md.external_params, param_name)
            value = parameters[string(param_name)]
            param_dims = parameter_dimensions(md, comp_name, param_name)
            num_dims = length(param_dims)

            set_external_param!(md, param_name, value; param_dims = param_dims, num_dims = num_dims)
```

Reply:
I see... I think having the one optional arg (`comp_name`) isn't so bad. If it's not provided, we can look for the param name using `unconnected_params`. (Might write a function for that so we can use it elsewhere.)

Reply:
I don't that that is enforced. At some point, we might want to revisit the treatment of external vars...

Reply:
Agreed, I think `set_leftover_params()` itself is on our backlog list of things to revisit, but for now this change will simplify code and work.